README.md
<img src="https://avatars2.githubusercontent.com/u/514566?v=3&u=4615dfc4970d93dea5d3eaf996b7903ee6e24e20&s=140" align="right" />
---
# php_zopfli

![Logo of php_zopfli](docs/logo-large.png)
PHP-Extension for compressing content with `Zopfli` a good zlib (gzip, deflate) compatible compressor which produces smallest possible results.

| [![Build Status](https://travis-ci.org/clickalicious/php_zopfli.svg?branch=master)](https://travis-ci.org/clickalicious/php_zopfli) 	| [![Scrutinizer](https://img.shields.io/scrutinizer/g/clickalicious/php_zopfli.svg)](https://scrutinizer-ci.com/g/clickalicious/php_zopfli/) 	| [![clickalicious premium](https://img.shields.io/badge/clickalicious-premium-green.svg?style=flat)](https://www.clickalicious.de/) 	| [![Apache 2.0](https://img.shields.io/hexpm/l/plug.svg)](http://opensource.org/licenses/Apache-2.0) 	|
|---	|---	|---	|---	|
| [![GitHub issues](https://img.shields.io/github/issues/clickalicious/phpmemadmin.svg?style=flat)](https://github.com/clickalicious/php_zopfli/issues) 	| [![Coverage Status](https://coveralls.io/repos/clickalicious/php_zopfli/badge.svg)](https://coveralls.io/r/clickalicious/php_zopfli)  	| [![GitHub release](https://img.shields.io/github/release/clickalicious/php_zopfli.svg?style=flat)](https://github.com/clickalicious/php_zopfli/releases) 	| [![GitHub stars](https://img.shields.io/github/stars/clickalicious/php_zopfli.svg?style=flat)](https://github.com/clickalicious/php_zopfli/stargazers)  	|

## Features

 - Compress content with `Zopfli`
 - Better compression & smaller results - up to 5%!
 - Con: This compressor takes more time (~100x slower!) then average zlib compressing


## Philosophy

`php_zopfli` was designed as a compressor for cases where the time to compress doesn't matter but the results should be as small as possible. I decided to implement `Zopfli` compression for this special case. `Zopfli` produces **up to 5% smaller results** than average zlib compressors out in the wild. `Zopfli` is **up to 100x times slower** to compress the data.


## Installation

Put this single line in your php.ini and put the extension in the same place as the other default extensions
```
extension=php_zopfli.(so|dll)
```

## Config

Optional configure the default number of iterations in your php.ini. Whenever you call zopfliencode() without the 2nd argument the php.ini default value is used instead.
```
[Zopfli]
zopfli.iterations = 5
```


## Example
`php_zopfli` can be used as replacement for *gzip/zlib*-compression. The `Zopfli` compression is **100% compatible with gzip + zlib** so your browser will understand the `Zopfli` output already. Just call *zopfliencode()* instead of *gzencode()*. 

The second argument is the optional iteration counter (a value between 1 and [max?]1000).


```php
<?php

$buffer = 'put some longer story here ...';

/*
 * Use zopfliencode() as replacement for gzencode()
 * 
 * @param string  The string to encode/compress
 * @param integer Number of iterations for compression (1 - 1000 [default = 5])
 */
$data = zopfliencode($buffer, 1);

header('Content-Encoding: gzip');
header('Content-Length: '.strlen($data));
header('Cache-Control: no-cache, no-store, max-age=0, must-revalidate');
header('Pragma: no-cache');

// output data
echo $data;

?>
```


## Documentation

There is currently no documentation.


## Versioning
For a consistent versioning i decided to make use of `Semantic Versioning 2.0.0` http://semver.org. Its easy to understand, very common and known from many other software projects.


## Roadmap

- [ ] General stability improvements and tests


## More Information
`Zopfli` was written by Jyrki Alakuijala, Ph.D. and Lode Vandevenne, M.Sc. from Google Inc.
You will find the projects homepage [here](http://code.google.com/p/zopfli/ "Go to Zopfli homepage") and the whitepaper [here](http://code.google.com/p/zopfli/downloads/detail?name=Data_compression_using_Zopfli.pdf "Download Zopfli whitepaper").


## Participate & share

... yeah. If you're a code monkey too - maybe we can build a force ;) If you would like to participate in either **Code**, **Comments**, **Documentation**, **Wiki**, **Bug-Reports**, **Unit-Tests**, **Bug-Fixes**, **Feedback** and/or **Critic** then please let me know as well!
<a href="https://twitter.com/intent/tweet?hashtags=&original_referer=http%3A%2F%2Fgithub.com%2F&text=%23php_zopfli%20-%20PHP-Extension%20for%20compressing%20data%20with%20%23Zopfli:%20https%3A%2F%2Fgithub.com%2Fclickalicious%2Fphp_zopfli&tw_p=tweetbutton" target="_blank">
  <img src="http://jpillora.com/github-twitter-button/img/tweet.png"></img>
</a>


## Sponsors  
Thanks to our sponsors and supporters:  

| JetBrains | Navicat |
|---|---|
| <a href="https://www.jetbrains.com/phpstorm/" title="PHP IDE :: JetBrains PhpStorm" target="_blank"><img src="https://www.jetbrains.com/phpstorm/documentation/docs/logo_phpstorm.png"></img></a> | <a href="http://www.navicat.com/" title="Navicat GUI - DB GUI-Admin-Tool for MySQL, MariaDB, SQL Server, SQLite, Oracle & PostgreSQL" target="_blank"><img src="http://upload.wikimedia.org/wikipedia/en/9/90/PremiumSoft_Navicat_Premium_Logo.png" height="55" /></a>  |


RELEASES
Version: 0.2.0
Date: 2015-03-01 15:00:00
Stability: stable

notes: 
        * Initial release
    





php_zopfli.c
/****************************************************************************
 *
 *  Copyright (c) 2013 - 2015 clickalicious GmbH (i.G.) - Benjamin Carl
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 ***************************************************************************/

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#ifdef linux
/* To enable CPU_ZERO and CPU_SET, etc. */
# define _GNU_SOURCE
#endif

#include "php.h"
#include "php_ini.h"
#include "ext/standard/info.h"
#include "php_zopfli.h"
#include "Zend/zend_extensions.h"

// Get zopfli required lib files
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Zopfli files required
#include "zopfli/gzip_container.h"
#include "zopfli/deflate.h"


/* compiled function list so Zend knows what's in this module */
zend_function_entry zopfli_functions[] = {
   ZEND_FE(zopfliencode, NULL)
   {NULL, NULL, NULL}
};

/* compiled module information */
zend_module_entry zopfli_module_entry = {
  #if ZEND_MODULE_API_NO >= 20010901
    STANDARD_MODULE_HEADER,           // Roughly means if PHP Version > 4.2.0
  #endif
    PHP_ZOPFLI_EXTNAME,               // Define PHP extension name
    zopfli_functions,                 /* Functions */
    NULL,                             /* MINIT */
    NULL,                             /* MSHUTDOWN */
    NULL,                             /* RINIT */
    NULL,                             /* RSHUTDOWN */
    NULL,                             /* MINFO */
  #if ZEND_MODULE_API_NO >= 20010901
    PHP_ZOPFLI_VERSION,               // Roughly means if PHP Version > 4.2.0
  #endif
    STANDARD_MODULE_PROPERTIES
};
#ifdef COMPILE_DL_ZOPFLI
ZEND_GET_MODULE(zopfli)     // Common for all PHP extensions which are build as shared modules
#endif


ZEND_FUNCTION(zopfliencode);


/* ZOPFLI */
ZEND_FUNCTION(zopfliencode)
{
    char          *buffer;
    size_t         len_buffer;
    char          *output      = 0;
    size_t         len_output  = 0;
    int            iterations  = 1;
    ZopfliOptions  options;
    char          *result;

    if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "s|l", &buffer, &iterations) == FAILURE) {
        RETURN_STRING("Bad parameters!", 1);
        //RETURN_NULL();
    }

    // size of input buffer
    len_buffer = strlen(buffer);

    // set options defaults
    options.verbose            = 0;
    options.numiterations      = iterations;
    options.blocksplitting     = 1;
    options.blocksplittinglast = 0;
    options.blocksplittingmax  = 15;

    // compress input
    ZopfliGzipCompress(&options, buffer, len_buffer, &output, &len_output);

    // return result STRINGL ->
    // cause otherwise binary data could be cutted randomly at \0 null byte in string
    RETURN_STRINGL(output, len_output, 1);
}


php_zopfli.h
/****************************************************************************
 *
 *  Copyright (c) 2013 - 2015 clickalicious GmbH (i.G.) - Benjamin Carl
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 ***************************************************************************/

#ifndef PHP_ZOPFLI_H
#define PHP_ZOPFLI_H 1

extern zend_module_entry zopfli_module_entry;
#define phpext_zopfli_ptr &zopfli_module_entry

#define PHP_ZOPFLI_VERSION "0.2.0"
#define PHP_ZOPFLI_EXTNAME "zopfli"

#ifdef PHP_WIN32
#define PHP_ZOPFLI_API __declspec(dllexport)
#include "win32/php_zopfli_win32.h"
#include "win32/php_zopfli_win32.c"
#else
#define PHP_ZOPFLI_API
#endif


#ifdef ZTS
#include "TSRM.h"
#endif

/**
 * PHP default module methods
 */
PHP_MINIT_FUNCTION(zopfli);
PHP_MSHUTDOWN_FUNCTION(zopfli);
PHP_RINIT_FUNCTION(zopfli);
PHP_RSHUTDOWN_FUNCTION(zopfli);
PHP_MINFO_FUNCTION(zopfli);

/**
 * declare methods available in PHP
 */
PHP_FUNCTION(zopfliencode);

#endif  /* PHP_ZOPFLI_H */


win32/php_zopfli_win32.c
/****************************************************************************
 *
 *  Copyright (c) 2013 clickalicious UG
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 ***************************************************************************/

int getrusage(int who, struct rusage * rusage)
{
    FILETIME starttime;
    FILETIME exittime;
    FILETIME kerneltime;
    FILETIME usertime;
    ULARGE_INTEGER li;

    if (who != RUSAGE_SELF) {
        // Only RUSAGE_SELF is supported in this implementation for now
        errno = EINVAL;
        return -1;
    }

    if (rusage == (struct rusage *) NULL)
    {
        errno = EFAULT;
        return -1;
    }

    memset(rusage, 0, sizeof(struct rusage));

    if (GetProcessTimes(GetCurrentProcess(),
                        &starttime, &exittime, &kerneltime, &usertime) == 0)
    {
        return -1;
    }

    /* Convert FILETIMEs (0.1 us) to struct timeval */
    memcpy(&li, &kerneltime, sizeof(FILETIME));
    li.QuadPart /= 10L;         /* Convert to microseconds */
    rusage->ru_stime.tv_sec = (long)li.QuadPart / 1000000L;
    rusage->ru_stime.tv_usec = (long)li.QuadPart % 1000000L;

    memcpy(&li, &usertime, sizeof(FILETIME));
    li.QuadPart /= 10L;         /* Convert to microseconds */
    rusage->ru_utime.tv_sec = (long)li.QuadPart / 1000000L;
    rusage->ru_utime.tv_usec = (long)li.QuadPart % 1000000L;


    // success
    return 0;
}


win32/php_zopfli_win32.h
/****************************************************************************
 *
 *  Copyright (c) 2013 clickalicious UG
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 ***************************************************************************/

#ifndef PHP_ZOPFLI_WIN32_H
#define PHP_ZOPFLI_WIN32_H


/****************************************************************************

    windows specific data-type, definitions, ...

****************************************************************************/
#include <windows.h>

#ifdef _MSC_VER
    typedef __int32 int32_t;
    typedef unsigned __int32 uint32_t;
    typedef __int64 int64_t;
    typedef unsigned __int64 uint64_t;
#else
    #include <stdint.h>
#endif

#if !defined(uint64)
    typedef unsigned __int64 uint64;
#define uint64 uint64
#endif


/****************************************************************************

  sysconf() replacement win32 = sysinfo()

****************************************************************************/
SYSTEM_INFO sysinfo;


/****************************************************************************

    cpu_set_t

****************************************************************************/
#if !defined(cpu_set_t)
    typedef unsigned long cpu_set_t;
#endif


/****************************************************************************

    rusage

****************************************************************************/
#ifdef WIN32
#include <time.h>               /* for struct timeval */
#endif
#ifndef WIN32
#include <sys/time.h>           /* for struct timeval */
#include <sys/times.h>          /* for struct tms */
#endif
#include <limits.h>             /* for CLK_TCK */

#define RUSAGE_SELF     0
#define RUSAGE_CHILDREN (-1)

struct rusage {
    struct timeval ru_utime;    /* user time used */
    struct timeval ru_stime;    /* system time used */
};

extern int getrusage(int who, struct rusage * rusage);


#endif  /* PHP_ZOPFLI_WIN32_H */


zopfli.c
/****************************************************************************
 *
 *  Copyright (c) 2013 clickalicious UG
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 ***************************************************************************/

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#ifdef linux
/* To enable CPU_ZERO and CPU_SET, etc. */
# define _GNU_SOURCE
#endif

// general extension related
#include "php.h"
#include "php_ini.h"
#include "ext/standard/info.h"
#include "php_zopfli.h"
#include "Zend/zend_extensions.h"

// get required lib files
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


// PHP ini mod
PHP_INI_BEGIN()
PHP_INI_ENTRY("zopfli.iterations", "5", PHP_INI_ALL, NULL)
PHP_INI_END()

// compiled function list so Zend knows what's in this module
zend_function_entry zopfli_functions[] = {    
   ZEND_FE(zopfliencode, NULL)
   {NULL, NULL, NULL}
};

/* compiled module information */    
zend_module_entry zopfli_module_entry = {
  #if ZEND_MODULE_API_NO >= 20010901
    STANDARD_MODULE_HEADER,           // Roughly means if PHP Version > 4.2.0
  #endif
    PHP_ZOPFLI_EXTNAME,               // Define PHP extension name
    zopfli_functions,                 /* Functions */
    PHP_MINIT(zopfli),                /* MINIT */
    PHP_MSHUTDOWN(zopfli),            /* MSHUTDOWN */
    NULL,                             /* RINIT */
    NULL,                             /* RSHUTDOWN */
    NULL,                             /* MINFO */
  #if ZEND_MODULE_API_NO >= 20010901
    PHP_ZOPFLI_VERSION,               // Roughly means if PHP Version > 4.2.0  
  #endif
    STANDARD_MODULE_PROPERTIES
};
#ifdef COMPILE_DL_ZOPFLI
ZEND_GET_MODULE(zopfli)               // Common for all PHP extensions which are build as shared modules
#endif

/* ? */
ZEND_FUNCTION(zopfliencode);

/* ZOPFLI */
ZEND_FUNCTION(zopfliencode)
{
    char    *buffer;
    size_t   len_buffer;
    char    *output      = 0;
    size_t   len_output  = 0;
    long     iterations;
    int      len_iterations;
    Options  options;

    if (zend_parse_parameters(
        ZEND_NUM_ARGS() TSRMLS_CC, 
        "s|l", 
        &buffer, 
        &len_buffer, 
        &iterations, 
        &len_iterations) == FAILURE
    ) {
        RETURN_STRING("Bad parameters!", 1);
    }

    if (!iterations) {
        iterations = INI_INT("zopfli.iterations");
    }

    // set options defaults
    options.verbose            = 0;
    options.numiterations      = (int)iterations;
    options.blocksplitting     = 1;
    options.blocksplittinglast = 0;
    options.blocksplittingmax  = 15;

    // compress input
    GzipCompress(
        &options, 
        buffer, 
        len_buffer, 
        &output, 
        &len_output
    );
    
    // return result STRINGL otherwise binary data could 
    // be cutted randomly at \0 null byte in string
    RETURN_STRINGL(output, len_output, 1);
}

// init method
PHP_MINIT_FUNCTION(zopfli)
{
    REGISTER_INI_ENTRIES();
    return SUCCESS;
}

// shutdown method
PHP_MSHUTDOWN_FUNCTION(zopfli)
{
    UNREGISTER_INI_ENTRIES();
    return SUCCESS;
}
